hibernate-core/src/main/java/org/hibernate/dialect/StructHelper.java

```diff
@@ -115,20 +114,58 @@ public static Object[] getJdbcValues(
 				attributeIndex = orderMapping[i];
 			}
 			jdbcIndex += injectJdbcValue(
-					embeddableMappingType.getAttributeMapping( attributeIndex ),
-					attributeValues,
+					getEmbeddedPart( embeddableMappingType, jdbcValueCount, attributeIndex ),
+					values,
 					attributeIndex,
 					jdbcValues,
 					jdbcIndex,
 					options
 			);
 		}
-		assert jdbcIndex == jdbcValueCount;
+		assert jdbcIndex == valueCount;
 		return jdbcValues;
 	}
 
+	public static Object[] getValues(EmbeddableMappingType embeddableMappingType, Object domainValue) {
+		final Object[] attributeValues = embeddableMappingType.getValues( domainValue );
+		if ( embeddableMappingType.isPolymorphic() ) {
+			final Object[] result = new Object[attributeValues.length + 1];
+			System.arraycopy( attributeValues, 0, result, 0, attributeValues.length );
+			result[attributeValues.length] = domainValue.getClass().getName();
```

Comment:
This "hack" could go away if we decide to add the FQN as last attribute value to the array.